src/matvis/gpu/beams.py

```diff
@@ -0,0 +1,178 @@
+
+        Parameters
+        ----------
+        tx, ty
+            Coordinates to evaluate the beam at, in sin-projection.
+        """
+        if self.use_interp:
+            self._interp(tx, ty, out)
+        else:
+            self._eval(self, tx.get(), ty.get(), self._np_beam)
+            out.set(self._np_beam)
+
+    def _interp(
+        self,
+        tx: cp.ndarray,
+        ty: cp.ndarray,
+        out: cp.ndarray,
+    ):
+        """Perform the beam interpolation, choosing between CPU and GPU as necessary."""
+        az, za = coordinates.enu_to_az_za(enu_e=tx, enu_n=ty, orientation="uvbeam")
+
+        # Set all the elements
+        self.interpolated_beam[..., len(az) :] = 0.0
+
+        gpu_beam_interpolation(
+            self.beam_data, self.daz, self.dza, az, za, beam_at_src=out
+        )
+
+
+def gpu_beam_interpolation(
+    beam: np.ndarray | cp.ndarray,
+    daz: float,
+    dza: float,
+    az: np.ndarray | cp.ndarray,
+    za: np.ndarray | cp.ndarray,
+    beam_at_src: cp.ndarray | None = None,
+):
+    """
+    Interpolate beam values from a regular az/za grid using GPU.
+
+    Parameters
+    ----------
+    beam
+        The beam values. The shape of this array should be
+        ``(nbeam, nax, nfeed, nza, naz)``. This is the axis ordering returned by
+        UVBeam.interp. This array can either be real or complex. Either way, the output
+        is complex.
+    daz, dza
+        The grid sizes in azimuth and zenith-angle respectively.
+    az, za
+        The azimuth and zenith-angle values of the sources to which to interpolate.
+        These should be  1D arrays. They are not treated as a "grid".
+
+    Returns
+    -------
+    beam_at_src
+        The beam interpolated at the sources. The shape of the array is
+        ``(nbeam, nfeed, nax, nsrc)``. The array is always complex (at single or
+        double precision, depending on the input).
+    """
+    beam = cp.asarray(beam)
+    az = cp.asarray(az)
+    za = cp.asarray(za)
+
+    # Get precision from the beam object.
+    if beam.dtype in (np.dtype("float32"), np.dtype("complex64")):
+        ctype = np.dtype("complex64")
+    elif beam.dtype in (np.dtype("float64"), np.dtype("complex128")):
+        ctype = np.dtype("complex128")
+    else:
+        raise ValueError(
+            f"Got {beam.dtype} as the dtype for beam, which is unrecognized"
+        )
+
+    complex_beam = beam.dtype.name.startswith("complex")
+    az /= daz
+    za /= dza
+
+    nbeam, nax, nfeed, nza, naz = beam.shape
+    nsrc = len(az)
+
+    if beam_at_src is None:
+        beam_at_src = cp.zeros((nbeam, nfeed, nax, nsrc), dtype=beam.dtype)
+    else:
+        assert beam_at_src.shape == (nbeam, nfeed, nax, nsrc)
+
+    for bm, fd, ax in itertools.product(range(nbeam), range(nfeed), range(nax)):
+        ndimage.map_coordinates(
+            beam[bm, ax, fd],
+            cp.asarray([za, az]),
+            order=1,
```

Comment:
Does this mean that we can do bicubic interpolation on GPU now as well? Do we want to add that option?

Reply:
yes it does! I will add that.